test/e2e/node_reuse.go

```diff
@@ -87,29 +86,24 @@ func nodeReuse(ctx context.Context, inputGetter func() NodeReuseInput) {
 
 	By("Download image")
 	imageURL, imageChecksum := ensureImage(upgradedK8sVersion)
-	By("Update KCP Metal3MachineTemplate with upgraded image to boot and set nodeReuse field to 'True'")
+	By("Set nodeReuse field to 'True' and create new KCP Metal3MachineTemplate with upgraded image to boot")
 	m3machineTemplateName := fmt.Sprintf("%s-controlplane", input.ClusterName)
 	updateNodeReuse(ctx, input.Namespace, true, m3machineTemplateName, managementClusterClient)
-	updateBootImage(ctx, input.Namespace, m3machineTemplateName, managementClusterClient, imageURL, imageChecksum, "raw", "md5")
+	newM3machineTemplateName := fmt.Sprintf("%s-new-controlplane", input.ClusterName)
+	createNewM3machineTemplate(ctx, input.Namespace, newM3machineTemplateName, m3machineTemplateName, managementClusterClient, imageURL, imageChecksum, "raw", "md5")
 
 	Byf("Update KCP to upgrade k8s version and binaries from %s to %s", kubernetesVersion, upgradedK8sVersion)
 	kcpObj := framework.GetKubeadmControlPlaneByCluster(ctx, framework.GetKubeadmControlPlaneByClusterInput{
 		Lister:      managementClusterClient,
 		ClusterName: input.ClusterName,
 		Namespace:   input.Namespace,
 	})
-	patch := []byte(fmt.Sprintf(`{
-		"spec": {
-			"rolloutStrategy": {
-				"rollingUpdate": {
-					"maxSurge": 0
-				}
-			},
-			"version": "%s"
-		}
-	}`, upgradedK8sVersion))
-	err := managementClusterClient.Patch(ctx, kcpObj, client.RawPatch(types.MergePatchType, patch))
-	Expect(err).To(BeNil(), "Failed to patch KubeadmControlPlane")
+	helper, err := patch.NewHelper(kcpObj, managementClusterClient)
+	Expect(err).NotTo(HaveOccurred())
+	kcpObj.Spec.MachineTemplate.InfrastructureRef.Name = newM3machineTemplateName
+	kcpObj.Spec.Version = upgradedK8sVersion
+	kcpObj.Spec.RolloutStrategy.RollingUpdate.MaxSurge.IntVal = 0
+	Expect(helper.Patch(ctx, kcpObj)).To(Succeed())
```

Comment:
This is not correct based on what I previously dealt with usage of new / old m3mts'. The problem is: here you already have KCP and 3 control plane machines under it created and 3 metal3 machines / 3 BMH one more layer below. Existing m3machines are created using existing m3mt. Now, you are creating a new m3mt here and pointing to it from the KCP, however controller logic to check the node reuse is set or not from the m3m point of view and from which m3mt it was created from. So, the logic would be: getting the exisitng m3m, find it is m3mt which it is created from and check for node reuse field. In your case, if you want your new m3mt to be used properly, just pointing from KCP to the new one is not enough. What you have to do to use it is, scale down the machines (as n consequence m3m which was using old m3mt being deleted) and scale up (as a consequence m3m would use new template you are pointing).

Reply:
Based on the offline call we had with @Sunnatillo, this should be fine to leave as is. Patching the same template should not hurt, but creating a new template is also fine and may be a better approach.